pkg/v1/remote/multi_write.go

```diff
@@ -0,0 +1,172 @@
+//
+// Current limitations:
+// - All refs must share the same repository.
+// - Images cannot consist of stream.Layers.
+func MultiWrite(m map[name.Reference]Taggable, options ...Option) error {
+	// Collect unique blobs (layers and config blobs).
+	blobs := map[v1.Hash]v1.Layer{}
+	for _, i := range m {
+		if img, ok := i.(v1.Image); ok {
+			if err := addImageBlobs(img, blobs); err != nil {
+				return err
+			}
+			continue
+		}
+		if idx, ok := i.(v1.ImageIndex); ok {
+			if err := addIndexBlobs(idx, blobs); err != nil {
+				return err
+			}
+			continue
+		}
+		return fmt.Errorf("pushable resource was not Image or ImageIndex: %T", i)
+	}
+
+	// Determine the repository being pushed to; if asked to push to
+	// multiple repositories, give up.
+	var repo, zero name.Repository
+	for ref := range m {
+		if repo == zero {
+			repo = ref.Context()
+		} else if ref.Context() != repo {
+			return fmt.Errorf("MultiWrite can only push to the same repository (saw %q and %q", repo, ref.Context())
+		}
+	}
+	o, err := makeOptions(repo, options...)
+	if err != nil {
+		return err
+	}
+	// Determine if any of the layers are Mountable, because if so we need
+	// to request Pull scope too.
+	ls := []v1.Layer{}
+	for _, l := range blobs {
+		ls = append(ls, l)
+	}
+	scopes := scopesForUploadingImage(repo, ls)
+	tr, err := transport.New(repo.Registry, o.auth, o.transport, scopes)
+	if err != nil {
+		return err
+	}
+	w := writer{
+		repo:    repo,
+		client:  &http.Client{Transport: tr},
+		context: o.context,
+	}
+
+	// Upload individual blobs in goroutines and collect any errors.
+	var g errgroup.Group
+	for _, b := range blobs {
+		b := b
+
+		// Handle foreign layers.
+		mt, err := b.MediaType()
+		if err != nil {
+			return err
+		}
+		if !mt.IsDistributable() {
+			// TODO(jonjohnsonjr): Add "allow-nondistributable-artifacts" option.
+			continue
+		}
+
+		g.Go(func() error { return w.uploadOne(b) })
+	}
+	if err := g.Wait(); err != nil {
+		return err
+	}
+
+	// With all of the constituent elements uploaded, upload the manifests
+	// to commit the images and indexes.
+	for ref, i := range m {
+		ref := ref
+		i := i
+		g.Go(func() error { return w.commitImage(i, ref) })
+	}
+	return g.Wait()
+}
+
+func addIndexBlobs(idx v1.ImageIndex, blobs map[v1.Hash]v1.Layer) error {
+	im, err := idx.IndexManifest()
+	if err != nil {
+		return err
+	}
+	for _, desc := range im.Manifests {
+		switch desc.MediaType {
+		case types.OCIImageIndex, types.DockerManifestList:
+			idx, err := idx.ImageIndex(desc.Digest)
+			if err != nil {
+				return err
+			}
+			if err := addIndexBlobs(idx, blobs); err != nil {
+				return err
+			}
+		case types.OCIManifestSchema1, types.DockerManifestSchema2:
+			img, err := idx.Image(desc.Digest)
+			if err != nil {
+				return err
+			}
+			if err := addImageBlobs(img, blobs); err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("unknown media type: %v", desc.MediaType)
```

Comment:
We need something like `v1.ImageIndex.Blob` to access non-manifest artifacts that are referenced by an index. Making the interface wider would be a breaking change, which is annoying, but I doubt it would affect much code outside of this repo...